File: seminars/01-dslib/ping-pong/retry.py
from dslib import Context, Message, Node


class PingClient(Node):
    def __init__(self, node_id: str, server_id: str):
        self._id = node_id
        self._server_id = server_id
        self._ping = None
        self._pong = None

    def on_local_message(self, msg: Message, ctx: Context):
        if msg.type == 'PING':
            self._ping = msg
            ctx.send(msg, self._server_id)
            ctx.set_timer('check_pong', 3)

    def on_message(self, msg: Message, sender: str, ctx: Context):
        # process messages from server
        if msg.type == 'PONG' and self._ping is not None:
            self._ping = None
            ctx.cancel_timer('check_pong')
            ctx.send_local(msg)

    def on_timer(self, timer_id: str, ctx: Context):
        # process fired timers here
        if timer_id == 'check_pong' and self._ping is not None:
            ctx.send(self._ping, self._server_id)
            ctx.set_timer('check_pong', 3)


class PingServer(Node):
    def __init__(self, node_id: str):
        self._id = node_id

    def on_local_message(self, msg: Message, ctx: Context):
        # not used
        pass

    def on_message(self, msg: Message, sender: str, ctx: Context):
        # process messages from client
        pong = Message('PONG', {'value': msg['value']})
        ctx.send(pong, sender)

    def on_timer(self, timer_id: str, ctx: Context):
        # process fired timers here
        pass

File: week-01/anysystem-intro/ping-pong/tests/src/solution/retry.rs
use std::rc::Rc;

use anysystem::{Context, Message, Process, ProcessState};

#[derive(Clone)]
pub struct Client {
    server: String,
    ping: Option<Message>,
}

impl Client {
    pub fn new(server: &str) -> Self {
        Self {
            server: server.to_string(),
            ping: None,
        }
    }
}

impl Process for Client {
    fn on_message(&mut self, msg: Message, _from: String, ctx: &mut Context) -> Result<(), String> {
        if msg.tip == "PONG" {
            self.ping = None;
            ctx.cancel_timer("check-pong");
            ctx.send_local(msg);
        }
        Ok(())
    }

    fn on_local_message(&mut self, msg: Message, ctx: &mut Context) -> Result<(), String> {
        if msg.tip == "PING" {
            self.ping = Some(msg.clone());
            ctx.send(msg, self.server.clone());
            ctx.set_timer("check-pong", 3.);
        }
        Ok(())
    }

    fn on_timer(&mut self, timer: String, ctx: &mut Context) -> Result<(), String> {
        if timer == "check-pong" && self.ping.is_some() {
            ctx.send(self.ping.as_ref().unwrap().clone(), self.server.clone());
            ctx.set_timer("check-pong", 3.);
        }
        Ok(())
    }

    fn state(&self) -> Result<std::rc::Rc<dyn ProcessState>, String> {
        Ok(Rc::new(self.ping.clone()))
    }

    fn set_state(&mut self, state: std::rc::Rc<dyn ProcessState>) -> Result<(), String> {
        self.ping
            .clone_from(state.downcast_rc::<Option<Message>>().unwrap().as_ref());
        Ok(())
    }
}

#[derive(Clone)]
pub struct Server {}

impl Server {
    pub fn new() -> Self {
        Self {}
    }
}

impl Process for Server {
    fn on_message(&mut self, msg: Message, from: String, ctx: &mut Context) -> Result<(), String> {
        if msg.tip == "PING" {
            let resp = Message::new("PONG".to_string(), msg.data);
            ctx.send(resp, from);
        }
        Ok(())
    }

    fn on_local_message(&mut self, _msg: Message, _ctx: &mut Context) -> Result<(), String> {
        Ok(())
    }

    fn on_timer(&mut self, _timer: String, _ctx: &mut Context) -> Result<(), String> {
        Ok(())
    }
}
